src/basics.md

```diff
@@ -1101,6 +1101,68 @@ fn main() {
 }
 ```
 
+[ex-error-chain-backtrace]: #ex-error-chain-backtrace
+<a name="ex-error-chain-backtrace"></a>
+## Obtain backtrace of complex error scenarios
+
+[![error-chain-badge]][error-chain] [![cat-rust-patterns-badge]][cat-rust-patterns]
+
+Extend errors by appending new errors via [`chain_err`]. It provides a
+better context to understand an error in a form of a backtrace. The
+below recipes attemps to deserialize the value `256` into a `u8`. An
+error will bubble up from Serde then csv and finaly up to the user
+code.
+
+```rust
+# extern crate csv;
+#[macro_use]
+extern crate error_chain;
+# #[macro_use]
+# extern crate serde_derive;
+#
+# error_chain! {
+#     foreign_links {
+#         Reader(csv::Error);
+#     }
+# }
+
+#[derive(Debug, Deserialize)]
+struct Rgb {
+    red: u8,
+    blue: u8,
+    green: u8,
+}
+
+fn read_rgb_from_csv(csv_data: &[u8]) -> Result<Rgb> {
+    let color: Rgb = csv::Reader::from_reader(csv_data)
+        .deserialize()
+        .nth(0)
+        .ok_or("Cannot deserialize the first CSV record")?
+        .chain_err(|| "Cannot deserialize RGB color")?;
+
+    Ok(color)
+}
+
+# type Hex = u32;
+#
+# fn rgb_to_hex(color: &Rgb) -> Hex {
+#     (color.red as Hex) << 16 | (color.blue as Hex) << 8 | (color.green as Hex)
+# }
+#
+fn run() -> Result<()> {
+    let csv = "red,blue,green
```

Comment:
to bump the size of our backtrace we can use a file input. Provide an associate function `from_path` which would in turn call associated function `from_reader` (formerly `read_rgb_from_csv`)

Reply:
Not sure I understand this comment. [`from_path`](https://docs.rs/csv/1.0.0-beta.4/csv/struct.ReaderBuilder.html#method.from_path) is already provided by the csv crate. 

The snippet would look like:

```rust
    let color: Rgb = csv::ReaderBuilder::new()
        .from_path("color.csv")
        .chain_err(|| "Cannot read CSV file")?
        .deserialize()
        .nth(0)
        .ok_or("Cannot deserialize the first CSV record")?
        .chain_err(|| "Cannot deserialize RGB color")?;
```

Then it leads to a shorter error stack:

```text
Error level - description
└> 0 - Cannot read CSV file
└> 1 - No such file or directory (os error 2)
└> 2 - No such file or directory (os error 2)
```

Am I missing something?

Reply:
Again my bad for lack of diligence in the review.

Reply:
That's alright. Maybe the following snippet will make the example more Rust idiomatic?

```rust
impl Rgb {
    fn from_reader(csv_data: &[u8]) -> Result<Rgb> {
        // move code of read_rgb_from_csv here
    }
}
```

and in the `run` function, the line will be:

```
let rgb = Rgb::from_reader(csv.as_bytes()).chain_err(|| "Cannot read CSV data")?;
```

The error stack is unchanged. What do you think?

Reply:
yep that's cool 👍 the comment slipped past me